google_guest_agent/network/systemd_networkd.go

```diff
@@ -0,0 +1,112 @@
+package network
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/run"
+	"github.com/GoogleCloudPlatform/guest-agent/metadata"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+const configDir = "/etc/systemd/network"
+
+type systemdnetworkd struct{}
+
+func (n systemdnetworkd) IsManaging(ctx context.Context, iface string) (bool, error) {
+	// First check if the service is running.
+	res := run.WithOutput(ctx, "/bin/sh", "-c", "systemctl status systemd-networkd.service | grep Active:")
+	if res.ExitCode != 0 {
+		return false, fmt.Errorf("failed to get status of systemd-networkd service: %v", res.StdErr)
+	}
+	if !strings.Contains(res.StdOut, "running") {
+		// Service is not running, so it can't be managing an interface.
+		return false, nil
+	}
+
+	// Check systemd network configuration.
+	res = run.WithOutput(ctx, "/bin/sh", "-c", "networkctl status | grep "+iface)
```

Comment:
To check if systemd-networkd is managing an interface please consider doing the following:

```
  res = run.WithOutput(ctx, "/bin/sh", "-c", fmt.Sprintf("networkctl list %s --json=short", iface))
  if res.ExitCode != 0 {
    return false, fmt.Errorf("failed to check systemd-networkd network status: %v", res.StdErr)
  }

  interfaceStatus := make(map[string]any)

  if err := json.Unmarshal([]byte(res.StdOut), &interfaceStatus); err != nil {
    return false, fmt.Errorf("failed to unmarshal interface status: %+v", err)
  }

  adminstrativeState, found := interfaceStatus["AdministrativeState"]
  if !found {
    return false, fmt.Errorf("could not determine interface state, AdministrativeState is not present")
  }

  return adminstrativeState == "configured", nil
```

That way we don't trust solely on the stdout format but in a machine readable format (json in this case). We also make sure that we are querying only the interface of interest (the primary one).

Reply:
Done.